fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InferPredicates.java

```diff
@@ -0,0 +1,118 @@
+
+import com.google.common.collect.Lists;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * infer additional predicates for `LogicalFilter` and `LogicalJoin`.
+ */
+public class InferPredicates extends DefaultPlanRewriter<JobContext> {
+    PredicatePropagation propagation = new PredicatePropagation();
+    PullUpPredicates pollUpPredicates = new PullUpPredicates();
+
+    /**
+     * The logic is as follows:
+     * 1. poll up bottom predicate then infer additional predicates
+     *   for example:
+     *   select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id
+     *   1. poll up bottom predicate
+     *      select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id and t.id = 1
+     *   2. infer
+     *      select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id and t.id = 1 and t2.id = 1
+     *   finally transformed sql:
+     *      select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id and t2.id = 1
+     * 2. put these predicates into `otherJoinConjuncts` , these predicates are processed in the next
+     *   round of predicate push-down
+     */
+    @Override
+    public Plan visitLogicalJoin(LogicalJoin<? extends Plan, ? extends Plan> join, JobContext context) {
+        join = (LogicalJoin<? extends Plan, ? extends Plan>) super.visit(join, context);
+        Plan left = join.left();
+        Plan right = join.right();
+        Set<Expression> expressions = getAllExpressions(left, right, join.getOnClauseCondition());
+        List<Expression> otherJoinConjuncts = Lists.newArrayList(join.getOtherJoinConjuncts());
+        switch (join.getJoinType()) {
+            case INNER_JOIN:
+            case CROSS_JOIN:
+            case LEFT_SEMI_JOIN:
+            case RIGHT_SEMI_JOIN:
+                otherJoinConjuncts.addAll(inferNewPredicate(left, expressions));
+                otherJoinConjuncts.addAll(inferNewPredicate(right, expressions));
+                break;
+            case LEFT_OUTER_JOIN:
+            case LEFT_ANTI_JOIN:
+                otherJoinConjuncts.addAll(inferNewPredicate(right, expressions));
+                break;
+            case RIGHT_OUTER_JOIN:
+            case RIGHT_ANTI_JOIN:
+                otherJoinConjuncts.addAll(inferNewPredicate(left, expressions));
+                break;
+            default:
+                return join;
+        }
+        return join.withOtherJoinConjuncts(otherJoinConjuncts);
+    }
+
+    /**
+     * reference `inferOn`
+     */
+    @Override
+    public Plan visitLogicalFilter(LogicalFilter<? extends Plan> filter, JobContext context) {
+        filter = (LogicalFilter<? extends Plan>) super.visit(filter, context);
+        Set<Expression> filterPredicates = filter.accept(pollUpPredicates, null);
+        Set<Expression> filterChildPredicates = filter.child(0).accept(pollUpPredicates, null);
```

Comment:
This code seem like every plan child has many repeated traverse the all child plan tree by the pollUpPredicates.
I think we should support invoke poll up and infer predicate which just process one level
```java
filter = (LogicalFilter<? extends Plan>) super.visit(filter, context);
Set<Expression> childPredicates = pollUpPredicate(filter); // just pull up one level child
Set<Expression> mergedPredicates = ImmutableSet.Builder()
    .addAll(filter.getPredicates())
    .addAll(childPredicates);
Set<Expression> inferredPredicates = propagation.infer(mergedPredicates);
if (!inferredPredicates.equals(filter.getPredicates()) {
    return new LogicalFilter(inferredPredicates, filter.child());
}
return filter;
```